src/witness_generator/machines/fixed_lookup_machine.rs

```diff
@@ -76,60 +74,82 @@ impl FixedLookup {
+                                // if the lhs is constant, it's a match iff the values match
+                                .map(|left| (left == *right).then_some(right))
+                                // if it's not constant, it's a match
+                                .unwrap_or_else(|| Some(right)),
+                            // if we do not know the lhs, it's a match
+                            Err(_) => Some(right),
+                        }
+                    })
+                    .collect::<Option<Vec<_>>>()
+            })
+            // deduplicate values
+            .fold(None, |acc, new_value| {
+                match acc {
+                    None => Some(Ok(new_value)),
+                    Some(Ok(value)) => {
+                        // if a second row matches
+                        if value == new_value {
+                            // if it has the same value, we didn't lose anything
+                            Some(Ok(value))
+                        } else {
+                            // TODO: if some of the new value coincides with the existing one, use that partial knowledge
+                            // For this, the state here should be more granular to cover each column separately
+                            Some(Err(()))
+                        }
+                    }
+                    // propagate the error. TODO: short circuit here
+                    e => e,
+                }
+            })
+            // if the accumulator is None, no row matched, which is an error
+            .ok_or_else(|| EvalError::Generic("Plookup is not satisfied".to_string()))?;
 
-        let rhs_evaluator =
-            ExpressionEvaluator::new(FixedEvaluator::new(fixed_data, rhs_row as usize));
+        let right_values = match res {
+            Ok(res) => res,
+            Err(()) => {
+                return Ok(vec![]);
+            }
+        };
 
         let mut reasons = vec![];
         let mut result = vec![];
-        for (l, r) in left.iter().zip(right.expressions.iter()).skip(1) {
+        for (l, r) in left.iter().zip(right_values) {
             match l {
                 Ok(l) => {
-                    // This needs to be a costant because symbolic variables
-                    // would reference a different row!
-                    let r = rhs_evaluator.evaluate(r).and_then(|r| {
-                        r.constant_value().ok_or_else(|| {
-                            format!("Constant value required: {}", r.format(fixed_data)).into()
-                        })
-                    });
-                    if let Err(err) = r {
-                        reasons.push(err);
-                        continue;
-                    }
-                    let r = r.unwrap();
                     let evaluated = l.clone() - r.clone().into();
                     // TODO we could use bit constraints here
                     match evaluated.solve() {
```

Comment:
If we change `solve()` to return a specific error for `is_invalid`, the case distinction here could also be simplified.

Reply:
I think I saw you made some changes to `solve` that in another PR, maybe this can be done in a different PR then?